### PR TITLE
[Enhancement] adjust cost model for right join of nestloopjoin

### PR DESCRIPTION
**Pull request by murphyatwork**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Review by Seaven — DISMISSED** [review]

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 14 / 14 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|12|12|100.00%|[]|

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

File: fe/fe-core/src/test/resources/sql/tpchcost/q7.sql [hunk 1]
```diff
@@ -45,12 +45,12 @@ Input Partition: UNPARTITIONED
 RESULT SINK
 
 25:MERGING-EXCHANGE
-cardinality: 250
+cardinality: 0
```
Comment (kangkaisen) [inline fe/fe-core/src/test/resources/sql/tpchcost/q7.sql:48]: Why the cardinality become zero?

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/EnumeratePlanTest.java [hunk 2]
```diff
@@ -167,7 +167,7 @@ public void testTPCHQ7EnumPlan() throws Exception {
                 "order by\n" +
                 "    o_year ;";
         int planCount = getPlanCount(sql);
-        Assert.assertEquals(51, planCount);
+        Assert.assertEquals(27, planCount);
```
Comment (kangkaisen) [inline fe/fe-core/src/test/java/com/starrocks/sql/plan/EnumeratePlanTest.java:170]: The tpch Q7 sql is wrong.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/CostModel.java [hunk 3]
```diff
@@ -301,14 +297,24 @@ public CostEstimate visitPhysicalMergeJoin(PhysicalMergeJoinOperator join, Expre
...
         public CostEstimate visitPhysicalNestLoopJoin(PhysicalNestLoopJoinOperator join, ExpressionContext context) {
-            final double nestLoopPunishment = 1000000.0;
             Statistics leftStatistics = context.getChildStatistics(0);
             Statistics rightStatistics = context.getChildStatistics(1);
 
             double leftSize = leftStatistics.getOutputSize(context.getChildOutputColumns(0));
             double rightSize = rightStatistics.getOutputSize(context.getChildOutputColumns(1));
-            return CostEstimate.of(leftSize * rightSize + nestLoopPunishment,
-                    rightSize * StatsConstants.CROSS_JOIN_COST_PENALTY * 2, 0);
+            double cpuCost = leftSize * rightSize + StatsConstants.CROSS_JOIN_COST_PENALTY;
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/CostModel.java:305]: the old way is multiply CROSS_JOIN_COST_PENALTY， not plus.  if the data size is very larger, plus the CROSS_JOIN_COST_PENALTY seems no effect.

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 14 / 14 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|12|12|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/CostModel.java (hunk 3 above)
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/CostModel.java:305]: HashJoin is `cpuCost = left_carnality + right_cardinality`
NestLoopJoin is `cpuCost = left_cardinality * right_cardinality + COST_PENALTY`

The multiply of cardinality should much larger than the plus ?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 13 / 13 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|12|12|100.00%|[]|

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/EnumeratePlanTest.java (hunk 2 above)
Comment (mofeiatwork) [reply fe/fe-core/src/test/java/com/starrocks/sql/plan/EnumeratePlanTest.java]: fixed

File: fe/fe-core/src/test/resources/sql/tpchcost/q7.sql (hunk 1 above)
Comment (mofeiatwork) [reply fe/fe-core/src/test/resources/sql/tpchcost/q7.sql]: fixed

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 14 / 14 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|12|12|100.00%|[]|

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Review by kangkaisen — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Comment by mofeiatwork** [comment]
@Mergifyio backport branch-2.4

**Pull request merged**